### PR TITLE
TOOL-658 #comment Updated to retry upload if failed.

### PR DESCRIPTION
**Pull request by richard-bogdan-bitrise**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -98,40 +100,55 @@ func main() {
...
-			if strings.TrimSpace(filePath) == "" {
-				continue
-			}
+		if err := uploadWithRetry(filelist, client, owner, repo, newRelease.GetID()); err != nil {
+			failf("Error during upload: %s", err)
+		}
+	}
+}
+
+func uploadWithRetry(filelist string, client *github.Client, owner string, repo string, id int64) error {
```
Comment (lpusok) [inline main.go]: NOTE: I would prefer to keep the parsing of the filelist in the main or an other function, to be more testable and create a function signature more clearly describing the functionality.

**Comment by lpusok** [comment]
Check my note.

File: main.go (hunk 1 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Divided code to more parts, please check.

File: main.go [hunk 2]
```diff
@@ -97,41 +103,81 @@ func main() {
...
+
+		if err := uploadFileWithRetry(asset.path, asset.displayFileName, fi, client, owner, repo, id); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func uploadFileWithRetry(filePath string, fileName string, fi *os.File, client *github.Client, owner string, repo string, id int64) error {
+	if err := retry.Times(3).Wait(5 * time.Second).Try(func(attempt uint) error {
```
Comment (lpusok) [inline main.go]: Could simplify to `return retry.Times...`; or move to uploadFileListWithRetry function

File: main.go [hunk 3]
```diff
@@ -97,41 +103,81 @@ func main() {
 	log.Infof("Release created:")
 	log.Printf(newRelease.GetHTMLURL())
 
-	if filelist := strings.TrimSpace(c.FilesToUpload); filelist != "" {
-		fmt.Println()
-		log.Infof("Uploading assets:")
+	filelist, err := parseFilesListConfig(c.FilesToUpload)
```
Comment (lpusok) [inline main.go]: Move parse login after stepconf.parse part.
NOTE: Could check that files exist before creating release.

File: main.go [hunk 4]
```diff
@@ -97,41 +103,81 @@ func main() {
...
+	return nil
+}
+
+func getFileNameFromPath(filePath string) (string, string, error) {
+	var fileName string
+	if s := strings.Split(filePath, "|"); len(s) > 1 {
+		if strings.TrimSpace(s[0]) != "" {
+			filePath = s[0]
+		} else {
+			return "", "", fmt.Errorf("Invalid file path configuration: %s", filePath)
```
Comment (lpusok) [inline main.go]: errors should begin with lowercase letter by Golang convention; check other occurences too.

File: main.go (hunk 2 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Simplified.

File: main.go (hunk 4 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Fixed.

File: main.go [hunk 5]
```diff
@@ -141,19 +160,19 @@ func uploadFileListWithRetry(assets []releaseAsset, client *github.Client, owner
...
+		if err := uploadFileWithRetry(GetUploader(uploadAsset, 3, 5000), asset.path, asset.displayFileName, fi, client, owner, repo, id); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func uploadFileWithRetry(filePath string, fileName string, fi *os.File, client *github.Client, owner string, repo string, id int64) error {
-	return retry.Times(3).Wait(5 * time.Second).Try(func(attempt uint) error {
+func uploadFileWithRetry(uploader *Uploader, filePath string, fileName string, fi *os.File, client *github.Client, owner string, repo string, id int64) error {
```
Comment (lpusok) [inline main.go]: Could add the number of retries as a return value, as matching on the log contents in the tests could break later.

File: main_test.go [hunk 6]
```diff
@@ -31,3 +38,30 @@ func TestParseRepo(t *testing.T) {
 		require.Equal(t, name, "steps-github-release")
 	}
 }
+
+func TestRetryUpload(t *testing.T) {
```
Comment (lpusok) [inline main_test.go]: We try to avoid adding new testify tests if possible.

**Pull request merged**